ENH: signal: added irrnotch and iirpeak functions.

antonior92 opened the pull request:
Hello, my name is Antonio and I am a Brazilian electrical engineer currently pursuing my master degree. This would be my first contribution to scipy. 

I have implemented and included two new functions to scipy: `iirnotch` and `iirpeak` which are
corespondent to the matlab functions with the same names ([iirnotch](http://www.mathworks.com/help/dsp/ref/iirnotch.html?s_tid=gn_loc_drop) and [iirpeak](http://www.mathworks.com/help/dsp/ref/iirpeak.html)).

The notch filter can be used to remove a frequency component from signal and can be useful, for instance, if you want to remove power-line interference from a signal. The peak filter is dual to the notch filter, attenuating every frequency component except the specified one.

Thanks for taking the time to consider my pull request.

nmayorov commented on scipy/signal/__init__.py:
```diff
@@ -64,6 +64,9 @@
    hilbert       -- Compute 1-D analytic signal, using the Hilbert transform.
    hilbert2      -- Compute 2-D analytic signal, using the Hilbert transform.
 
+   iirnotch      -- Design second-order IIR notch digital filter.
```
It should go to "Filter Design" section below or to "Matlab-style IIR filter design".

antonior92 replied:
I moved the both functions to "Matlab-style IIR filter design".

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3504,6 +3504,196 @@ def besselap(N, norm='phase'):
+    Parameters
+    ----------
+    w0 : float
+        A scalar that must satisfy  0 < w0 < 1. For notch filters,
```
Better to start the description from the definition, like "Normalized frequency to remove from the filtered signal". And then explain its properties.

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3504,6 +3504,196 @@ def besselap(N, norm='phase'):
+        w0 is the normalized frequency to remove from the filtered
+        signal (w0 = 1 corresponds to half of the sample frequency).
+    bw : float
+        Filter bandwidth.
```
Could it be explained better and in more detail? Maybe mention that at bandwidth edge frequencies the attenuation is determined by `ab`.

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3504,6 +3504,196 @@ def besselap(N, norm='phase'):
+    bw : float
+        Filter bandwidth.
+    ab : float, optional
+        Bandwidht maximum atenuation in decibels. By default, `ab`
```
Typos in "Bandwidht" and "atenuation".

I'm not really an expert, but "Bandwidht maximum atenuation in decibels" doesn't sound very clear. Is it written in some book? I would write it like "Attenuation in decibels at bandwidth edge frequencies" (it's just a guess, better to copy a phrase from some book).

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3504,6 +3504,196 @@ def besselap(N, norm='phase'):
+    b = gain*np.array([1.0, -2.0*np.cos(w0), 1.0])
+    a = np.array([1.0, -2.0*gain*np.cos(w0), (2.0*gain-1.0)])
+
+    return (b, a)
```
Parentheses are not necessary.

nmayorov commented:
If it to go to "Matlab-style IIR filter design" then function signatures are good. Otherwise, they can be tweaked somehow.

Someone who is more closely related to `scipy.signal` should take a look.

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -3504,6 +3504,196 @@ def besselap(N, norm='phase'):
+    >>> b, a = signal.iirnotch(w0, bw)
+
+    >>> # Frequency response
+    >>> w, h = signal.freqz(b, a, 2**14)
```
why 2 ** 14, that seems very high?

antonior92 replied:
The notch filter is a band-stop filter with very narrow bandwidth. For the exemple the magnitude of the filter frequency response will decrease very fast to zero around 60Hz.
If to few points are sampled from the frequency response the plot will not represent the frequency response properly. For instance, the graph below was generated for 2 ** 9 points.

![screenshot 2016-07-23 11 23 30](https://cloud.githubusercontent.com/assets/16557411/17078373/0ab1c5a0-50c8-11e6-8734-4688ac156bc4.png)

Eric89GXL replied:
If you lower the sample rate to say 200 Hz you could do fewer points,
right?

antonior92 replied:
Yes, for a lower sample rate fewer points would do. Do you suggest I do this?

Eric89GXL replied:
Yeah, it might be a bit better

antonior92 replied:
Ok!

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -3504,6 +3504,196 @@ def besselap(N, norm='phase'):
+    # Compute numerator b and denominator a
+    # formula 11.3.7 from reference [1]
+    b = gain*np.array([1.0, -2.0*np.cos(w0), 1.0])
+    a = np.array([1.0, -2.0*gain*np.cos(w0), (2.0*gain-1.0)])
```
So this method always produces a second-order filter, then? Please add a note to the docs.

antonior92 replied:
Yes this method always produce a second order filter.

Eric89GXL commented:
Thanks for looking into this, looks like a good start.

First, you need a rebase to latest `master`. (Don't merge -- use rebase.)

Can you add to the docs some comments on the relationship between this and band-stop IIR filter design? Also maybe add band-stop version to your example code.

We have started having `output=...` arguments, e.g. `'sos'` or `'zpk'` to our filter design functions, especially since high-order filters are often unstable with numerator/denominator coeffecient output. However, it looks like this only ever creates a second-order filter, is that correct? If so, could a high-order band-stop filter do better?

antonior92 commented on scipy/signal/filter_design.py:
```diff
@@ -3635,8 +3645,8 @@ def iirpeak(w0, bw, ab=3):
     --------
-    Design and plot filter to remove the 300Hz component from a
-    signal sampled at 1000Hz, using a quality factor Q=30
+    Design and plot filter to remove the frequêncies other than the 300Hz
```
I put a non-ascii caractere here by accident. The next commit will fix this.

antonior92 commented:
I would like to highlight some points about this methods:
* The `iirnotch` computes the coefficients of a band-stop filter with a narrow band and the `iirpeak` of a band-pass filter with narrow band.
* They always produce second order filters and therefore are less likely to suffer from numerical instability (I didn't include an `output=...` argument for this reason, but I can include it if you guys find that it would be better to have it).
*  If the application require a **single frequency** to be removed from the signal the notch filter have advantages in relation to a high-order band-stop filter because the notch filter is a low order filter
that manages to completely eliminate a frequency component from the signal (it includes a zero on the correspondent frequency that manages to do it) with a very narrow bandwidth.
On the other hand, if the application requires to reject a **range of frequencies** a band-stop filter would be preferable. A similar argument can be used to compare the peak filter with the band-pass filter.

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,206 @@ def besselap(N, norm='phase'):
+    Parameters
+    ----------
+    w0 : float
+        Normalized frequency to remove from filtered signal. It is a
```
"from a filtered signal" will read better.

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,206 @@ def besselap(N, norm='phase'):
+    w0 : float
+        Normalized frequency to remove from filtered signal. It is a
+        scalar that must satisfy  0 < w0 < 1, with w0 = 1 corresponding
+        to half of the sample frequency.
```
"Sampling frequency" sounds more conventional.

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,206 @@ def besselap(N, norm='phase'):
+
+    Examples
+    --------
+    Design and plot filter to remove the 60Hz component from a
```
I suggest to also plot a phase response to give an idea how this filter distorts a signal. Does it make sense?

antonior92 replied:
It make sense! I will add it on the next commit.

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,206 @@ def besselap(N, norm='phase'):
+        scalar that must satisfy  0 < w0 < 1, with w0 = 1 corresponding
+        to half of the sample frequency.
+    bw : float
+        Filter bandwidth (the diference between up and lower passband edges).
```
Perhaps mention that `bw` is also measured in the normalized units.

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,230 @@ def besselap(N, norm='phase'):
+    Parameters
+    ----------
+    w0 : float
+        Normalized frequency to be retained on the signal. It is a
```
Perhaps "to be retained in a signal"?

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,230 @@ def besselap(N, norm='phase'):
+        Normalized filter bandwidth (the diference between up and lower
+        passband edges).
+    ab : float, optional
+        Maximum allowed fluctuation of the logarithm gain (attenuation)
```
Sorry for being picky. But it seems to me that the word "fluctuation" doesn't fit here, because the gain curve looks simple with a single peak or notch (no fluctuations as such). 

Can you think of another definition or justify that this is the best?

antonior92 replied:
I changed it to "Maximum allowed deviation on the logarithm gain". Does it sounds better?

nmayorov commented:
My final suggestions:

1. Use the log-scale for frequencies in the plots.
2. Use degrees instead of radians for the phase.

I believe both are conventional choices for such plots.

Other than that, I think it's in a good shape. @Eric89GXL please make a final review and decision.

e-q commented:
For consistency's sake, instead of `ab`, I think it would be better to use `rp`/`rs` as is used in functions like `signal.ellip`. 

Instead of testing equivalence with hardcoded values from MATLAB, I'd prefer if the filters were tested more directly. I.e. verify via freqz (or even filtering) that the filter requirements are met.

Eric89GXL commented:
> Instead of testing equivalence with hardcoded values from MATLAB, I'd prefer if the filters were tested more directly. I.e. verify via freqz (or even filtering) that the filter requirements are met.

I would say replace *instead of* with *in addition to*, since it seems like both types of tests have value.

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,230 @@ def besselap(N, norm='phase'):
+    References
+    ----------
+    .. [1] Sophocles J. Orfanidis, "Introduction To Signal Processing",
+           Prentice-Hall, 1996
```
Do you happen to know the page number of the given edition? It helps people find it more easily.

Eric89GXL commented:
I like the example output now:

https://521-1460385-gh.circle-artifacts.com/0//home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.iirnotch.html#scipy.signal.iirnotch

But maybe it's worth showing a few different Q values?

antonior92 commented:
Ok so I am almost finishing the changes on the code and I would like to point out what suggestion I have fulfilled:

1. I have used degrees instead of radians.  
2. I have changed `ab` to `rp`.
3. I have added one test for each function. These test evaluate if the frequency response comply with the specifications.
4. I have include the page numbers.

I have not implement some of the suggestions and I would like to justify why:
* I have  not used log-scale for the frequencies. The reason is the following: while for analog filters the use of log-scale is ubiquitous, when work with digital filters that have a frequency response that goes from `0` to `pi` it is more commom to use linear scale (e.g. all the figures from [1] uses a linear scale when plotting normalised frequencies). Since we are working with a digital filters but choose to represent the frequencies using Hz we are kind of in a grey area and I think both would work. But in the case of notch/peak filters there are no great need to work with both small and large scales (that is what the log-scale does), the main interest is around the frequency `f0`, and the log-scale will only distort that region. 
* I have not showed different values of `Q` in the example. The reason is the following: I think the example plots would be more informative if I have used more than one value of `Q` and illustrated the effect of it,  nevertheless I think it would make the example more complicated. The way it is I think the user can copy and paste the example, adjust its parameters and would be able to use the filter out-of-the-box. If I started to add adicional analysis I think it would become less straight forward to do so. 

If the aforementioned comments did not manage to convince you guys I will implemented the suggestions when asked.


[1] Oppenheim, Alan V., and Ronald W. Schafer. Discrete-time signal processing. Pearson Higher Education, 2010.

nmayorov commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,232 @@ def besselap(N, norm='phase'):
+    bw : float
+        Normalized filter bandwidth (the diference between up and lower
+        passband edges).
+    rp : float, optional
```
I think it should be called `rs`. According to http://scipy.github.io/devdocs/generated/scipy.signal.iirfilter.html#scipy.signal.iirfilter: "For Chebyshev and elliptic filters, provides **the minimum attenuation in the stop band.** (dB)".

By the way let's use the same definition: "The minimum attenuation in the stop band (determined by `bw`) expressed in dB. Default is 3." Does it sound good?

nmayorov replied:
From MATLAB's `ellip` documentation: "... and Rs decibels of stopband attenuation down from the peak passband value." So I think `rs` is indeed a good and consistent name here.

antonior92 replied:
Nikolay I still think it is `rp`... Lets me explain my way of thinking and you check if I am wrong:

The figure bellows show the specifications of a band-stop filter:
![fullsizerender 2](https://cloud.githubusercontent.com/assets/16557411/17193152/11bd045a-5428-11e6-9f39-4f531f5a0c3c.jpg)


My interpretation is that the notch filter have an rejection band of a single point `w0` and an complete attenuation at that point `rs = infinity`

![fullsizerender 5](https://cloud.githubusercontent.com/assets/16557411/17193216/7524c316-5428-11e6-8c8e-1019f093448e.jpg)

nmayorov replied:
That is a good explanation. I think you are correct that the stopband in our case is a single point (I didn't think about it this way). So yes, let's keep `rp` then.

About the definition of `rp` (former `ab`). I thought of it more like the strength of the filter's attenuation (is it partially correct thinking?), but now I realize that it can (and probably should) be thought as passband sort of ripple (or fluctuations). So your definition makes more sense to me now, I think it's good.

To further clarify for myself I'm going to plot amplitude responses with different `rp` values.

nmayorov commented:
Well explained about a log-scale for frequencies, let's leave it in a linear scale.

As for the phase plot for `iirnotch`: wouldn't it be better to plot it from 0 to -360 so it looks continuous?

nmayorov commented:
An idea about Q-factor: if it such a convenient parameter it can be allowed as an alternative to `bw` as (for example) ``iirpeak(w0, bw=None, ab=3, Q=None)`` with an explanation that either `bw` or `Q` must be set. Is it known how to compute `bw` from `Q` with `ab != 3`? 

Alternatively, `bw` can be replaced by `Q` completely, like `iirpeak(w0, Q=10, ab=3)`.

But I'm not sure that it will be better than the current version, just an idea.

antonior92 commented:
About using Q as an alternative parameter: the problem is that I don't know how to compute Q for `ab != 3`. And I am afraid it may make things more confusing since each parameter would impose some restrictions on the others...

About the angle of `iirnotch` the discontinuity at `w0` is of `180` and reflects that the module goes from positive to negative at that frequency. The discontinuity could go from -90 to +90 or could go from -90 to -270, but it would be there nevertheless. I am already using unwrap so I don't know how could I restrain the fase from 0 to -360 without manually subtracting 360... What do you think?

nmayorov commented:
1. About Q-factor. I also suspected that it can bring unnecessary complications.
2. About plots. Sorry, I didn't realize that it was 180 degrees jump (only briefly looked at plots with radian scale). Sure nothing can be done with it.

e-q commented:
This isn't a hard objection, but  here are my two cents about `Q` vs `bw`: 

I use these kinds of filters often, and prefer using `Q`, since it's invariant in a geometric way. In effect, there is some degeneracy in `bw` and `rp`, i.e. in any notch produced in this way, there is always a band where the attenuation is -3dB, so one could argue for a simplified call signature of just `iirnotch(w0, Q)`. (This is what I use in some personal functions)

antonior92 commented:
Hey @e-q thank you for your comment. Your argument that use `Q` is better than using `bw` is
convincing and I agree that there are advantages in using `iirnotch(w0, Q)` as a signature. 

On the other hand, using `iirnotch(w0, bw)` as a signature the user can easily specify the `Q` factor
using the formula `bw = w0/Q`, clearly stated at the documentation, and have the extra liberty to set the parameter `rp`, calling the function as `iirnotch(w0, bw, rp)`.

nmayorov commented:
I would prefer `iirpeak(w0, Q)` version, unless it is explained what is the advantage of using `rp` and why it is helpful for this kind of filter (otherwise it looks like a parameter no one going to use).

Eric89GXL commented:
My sense is that Q is more natural as well

antonior92 commented:
Ok! I will change it.

antonior92 commented:
About the implementation, I would like your guys suggestions about something:
So far I have put the functions on `Matlab-style IIR filter design` but with the signature
change I don't think it fits there any longer. So I see two options:
1) Delete it from this place and change it to the filter design section.
2) Maintain functions on `Matlab-style IIR filter design` section with a signature 
identical to the matlab one and add functions `notchfilter(w0, Q)` and `peakfilter(w0, Q)` 
under the filter design section.
What do you guys think? I like the second option better, but I not sure about it.

Eric89GXL commented:
I think it still fits within the MATLAB-style section even if the function signature is not identical. If it can perform the same functionality with a similar signature it seems okay to put it there.

antonior92 commented:
I changed the signature to `irrnotch(w0, Q)` and `irrpeak(w0, Q)`, I was reluctant at first but I think it  looks nice!

antonior92 commented on scipy/signal/filter_design.py:
```diff
@@ -3653,27 +3653,26 @@ def besselap(N, norm='phase'):
+    Q : float
+        Quality factor. Dimensionless parameter that characterizes
+        notch filter bandwidth `bw` relative to its center
+        frequency, `Q = w0/bw`.
```
What do you guys think about this definition? I'm not sure about it...

antonior92 replied:
What do you guys think about this definition? I'm not sure about it...

Eric89GXL replied:
That looks good. You need double-backticks, though, since this is meant to be code, not a link.

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -3653,27 +3653,26 @@ def besselap(N, norm='phase'):
     (high quality factor). It rejects a narrow frequency band and
     leaves the rest of the spectrum little changed.
 
+    .. versionadded: 0.19.0
```
I think this usually goes in `Notes`

e-q commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,222 @@ def besselap(N, norm='phase'):
+    Q : float
+        Quality factor. Dimensionless parameter that characterizes
+        peak filter bandwidth ``bw`` relative to its center
+        frequency, ``Q = w0/bw``.
```
It may be helpful here to mention that ``bw`` is the 3dB bandwidth, now that `rp` is effectively fixed there.

antonior92 replied:
ok! I added this information on the docstring.

e-q commented:
This is looking good. Unfortunately, it looks like there are conflicts with the master branch. If you are able, you should try to rebase this PR off of the current state of the master branch.

antonior92 commented:
I rebase my branch. I think it is ok now.

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,222 @@ def besselap(N, norm='phase'):
+    beta = (gb/np.sqrt(1.0-gb**2.0))*np.tan(bw/2.0)
+
+    # Compute gain: formula 11.3.20 (p.579) from reference [1]
+    gain = 1.0/(1.0+beta)
```
These first 10 lines or so are identical, right? Could they be moved to a private function like `gain, w0 = _compute_notch_peak(w0, Q)`?

antonior92 replied:
Actually the beta is a little different for each case. But I could do something like:
``bw, gb, w0 = _compute_notch_peak(w0, Q)``. Do you think I should do it?

Eric89GXL replied:
so the only lines that differ are the `beta = ` and `a = ` lines in these two functions? If so, how about a private `return _compute_notch_peak(w0, Q, kind='notch' | 'peak')` or so?

Eric89GXL replied:
Or `b =` lines differ, rather

antonior92 replied:
Ok it sounds nice... So the private function signature will be like 
``b, a = _compute_notch_peak(w0, Q, kind='notch' | 'peak')``

Eric89GXL replied:
Yeah, and the public functions are basically one-liner wrappers to this function

antonior92 replied:
Ok!

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -3855,16 +3856,30 @@ def iirpeak(w0, Q):
-    # Compute gain: formula 11.3.20 (p.579) from reference [1]
-    gain = 1.0/(1.0+beta)
+        # Compute gain: formula 11.3.6 (p.575) from reference [1]
+        gain = 1.0/(1.0+beta)
```
To make it clearer which things are different, why not have a second conditional like:
```
if ftype == 'notch':
    beta = ...
elif ...
else ...
gain = 1.0/(1.0 + beta)
a = np.array([1, -2*gain*np.cos(w0), 2*gain - 1
if ftype == 'notch':
    b = ...
else:
    b = ...
return b, a
...
```
WDYT?

antonior92 replied:
I added this second conditional in the new commit. I think it's better, thanks!

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -3653,6 +3653,235 @@ def besselap(N, norm='phase'):
+        # Compute numerator b and denominator a
+        # formula 11.3.21 (p.579) from reference [1]
+        b = (1.0-gain)*np.array([1.0, 0.0, -1.0])
+        a = np.array([1.0, -2.0*gain*np.cos(w0), (2.0*gain-1.0)])
```
`a` is the same in both cases and thus can move outside the conditional

antonior92 replied:
ok!

Eric89GXL commented:
Other than my last gripe, LGTM +1 for merge.

@e-q @nmayorov are you happy now, too?

nmayorov commented:
+1 from me too.

e-q commented:
Looks good! Could probably use a squash before merging.

antonior92 commented:
Ok @e-q I squash all commits into one.

Eric89GXL commented:
Great, unless anyone else wants to have a look, I'll merge in a day or two.

antonior92 commented:
Hey guys, I rebase the branch to solve the conflicts that have shown up recently.

e-q commented:
In it goes! Thanks @antonior92
